### PR TITLE
[BOLT][BAT] Fix translate for branches added by BOLT

### PR DESCRIPTION
**Pull request by aaupov**
Map branches that are not present in BAT to the containing basic block.
The normal behavior is to map it to the preceding translation entry,
which may or may not be a basic block, and this causes profile staleness
issues.

This diff fulfills the intent of the comment added in 21f4303bfd01d:
```
// Branch source addresses are translated to the first instruction of the
// source BB to avoid accounting for modifications BOLT may have made in the
// BB regarding deletion/addition of instructions.
```

Test Plan: Updated bolt-address-translation-yaml.test

### CODE REVIEW

File: bolt/test/X86/bolt-address-translation-yaml.test [hunk 1]
```diff
@@ -24,6 +24,32 @@ READ-BAT-CHECK-NOT: BOLT-ERROR: unable to save profile in YAML format for input
...
+OBJDUMP-CHECK: jmp
+# Confirming it's not in BAT
+RUN: llvm-bat-dump %t.out --dump-all | FileCheck %s --check-prefix BAT-DUMP-CHECK
+BAT-DUMP-CHECK: Function Address: 0x800040, hash: 0x99e67ed32a203023
+# Containing basic block
+BAT-DUMP-CHECK: 0x34 -> 0x32 hash: 0x6c36179f229b0032
+# Branch entry just above
+BAT-DUMP-CHECK-NEXT: 0x37 -> 0x35 (branch)
+# No entry for that offset
+BAT-DUMP-NOT: 0x3b ->
```
Comment (maksfb) [inline bolt/test/X86/bolt-address-translation-yaml.test:42]: Is this line intended to be used? Is it needed?

File: bolt/tools/bat-dump/bat-dump.cpp [hunk 2]
```diff
@@ -48,10 +48,9 @@ static cl::opt<std::string> InputFilename(cl::Positional,
                                           cl::Required,
                                           cl::cat(BatDumpCategory));
 
-static cl::list<uint64_t> Translate("translate",
-                                    cl::desc("translate addresses using BAT"),
-                                    cl::value_desc("addr"),
-                                    cl::cat(BatDumpCategory));
+static cl::list<std::string>
+    Translate("translate", cl::desc("translate addresses using BAT"),
+              cl::value_desc("addr[:is_from]"), cl::cat(BatDumpCategory));
```
Comment (maksfb) [inline bolt/tools/bat-dump/bat-dump.cpp:53]:
```suggestion
              cl::value_desc("addr[:is_src]"), cl::cat(BatDumpCategory));
```

File: bolt/lib/Profile/BoltAddressTranslation.cpp [hunk 3]
```diff
@@ -478,18 +478,34 @@ uint64_t BoltAddressTranslation::translate(uint64_t FuncAddress,
...
 
   --KeyVal;
 
   const uint32_t Val = KeyVal->second >> 1; // dropping BRANCHENTRY bit
-  // Branch source addresses are translated to the first instruction of the
-  // source BB to avoid accounting for modifications BOLT may have made in the
-  // BB regarding deletion/addition of instructions.
-  if (IsBranchSrc)
-    return Val;
+  if (IsBranchSrc) {
```
Comment (maksfb) [inline bolt/lib/Profile/BoltAddressTranslation.cpp:494]: Can we handle the exact match case (from above) here as well to simplify the code?

File: bolt/test/X86/bolt-address-translation-yaml.test (hunk 1 above)
Comment (aaupov) [reply bolt/test/X86/bolt-address-translation-yaml.test:42]: Yes, it's a negative check that the branch is not in BAT.

File: bolt/lib/Profile/BoltAddressTranslation.cpp (hunk 3 above)
Comment (aaupov) [reply bolt/lib/Profile/BoltAddressTranslation.cpp:494]: No as we use `find` for exact match, and `upper_bound` for regular translation.

Comment (aaupov) [reply bolt/lib/Profile/BoltAddressTranslation.cpp:494]: Agree we can do an upper_bound followed by a comparison.

File: bolt/lib/Profile/BoltAddressTranslation.cpp [hunk 4]
```diff
@@ -492,6 +486,10 @@ uint64_t BoltAddressTranslation::translate(uint64_t FuncAddress,
 
   const uint32_t Val = KeyVal->second >> 1; // dropping BRANCHENTRY bit
   if (IsBranchSrc) {
+    // Branch entry is found in BAT
+    if (KeyVal->first == Offset && KeyVal->second & BRANCHENTRY)
```
Comment (maksfb) [inline bolt/lib/Profile/BoltAddressTranslation.cpp:490]: What would be the case where we need to check for the second condition (`KeyVal->second & BRANCHENTRY`)? Can we drop it?

**Pull request closed without merging**